### ISSUE TITLE
Downloading the csv for a distribution creates duplicate data

### ISSUE THREAD

**Issue by hverdonk** [vsoch/freegenes#109]
**Describe the bug**
When I download the csv file for a distribution, it lists multiple copies of the same well and its contents. I suspect the software is pulling the plate map from every plate we have in stock and appending it to the same file.

Also @Koeng101, the pOpen_v3 spreadsheet doesn't contain the plasmid sequence. Is it not in the database?

**To Reproduce**
Download the csv from either distribution and examine it. The problem is most evident for pOpen_v3, but can also be seen in OpenEnzyme_1.

**Expected behavior**
One plate map with no repeated data.

Here are copies of the files in question, so you don't have to go grab them:
[freegenes-distribution-popen_v3-individual-2019-10-31.xlsx](https://github.com/vsoch/freegenes/files/3796157/freegenes-distribution-popen_v3-individual-2019-10-31.xlsx)
[freegenes-distribution-openenzyme_1-2019-10-30.xlsx](https://github.com/vsoch/freegenes/files/3796158/freegenes-distribution-openenzyme_1-2019-10-30.xlsx)

**Comment by vsoch** [comment]
@Koeng101 is it possible that on one of the erroneous imports that duplicate wells were created? I remember we discussed this right before you started setting up testing (and this is my intuition).

**Comment by Koeng101** [comment]
Actually it isn’t! pOpen_v3 was an old distribution that was manually created. 

I think what Hannah is describing is accurate - it’s pulling from every plate associated with that distribution. Most likely, there is a `for plate in distributions.platemaps.plates` (didn’t want to write our two for loops, you get it), meanwhile a [platemap.plates[0] for platemap in distributions.platemaps] would be more appropriate for grabbing this data, since we only want 1 representative plate from each platemap

**Comment by vsoch** [comment]
okay, what I'll do is walk through this example. I don't think the issue is plates here because we still are representing the same data from the previous API, so there perhaps was an issue there? Let's walk through this in detail. First we will grab the plateset. This is identical locally and on the server. Here is the plateset:

```python
from fg.apps.main.models import *                                                                             
...
plateset                                                                                                      
 <PlateSet:pOpen_v3,18>

plateset.plates.count()                                                                                       
18
```
And there are in fact 18 plates:

```
for plate in plateset.plates.all():
    print(plate)

<Plate:<Container:FreeGenes_Rack_C>,pOpen_v3>
<Plate:<Container:FreeGenes_Rack_C>,pOpen_v3>
<Plate:<Container:FreeGenes_Rack_C>,pOpen_v3>
<Plate:<Container:FreeGenes_Rack_C>,pOpen_v3>
<Plate:<Container:FreeGenes_Rack_C>,pOpen_v3>
<Plate:<Container:FreeGenes_Rack_C>,pOpen_v3>
<Plate:<Container:FreeGenes_Rack_C>,pOpen_v3>
<Plate:<Container:FreeGenes_Rack_C>,pOpen_v3>
<Plate:<Container:FreeGenes_Rack_C>,pOpen_v3>
<Plate:<Container:FreeGenes_Rack_C>,pOpen_v3>
<Plate:<Container:FreeGenes_Rack_C>,pOpen_v3>
<Plate:<Container:FreeGenes_Rack_C>,pOpen_v3>
<Plate:<Container:FreeGenes_Rack_C>,pOpen_v3>
<Plate:<Container:FreeGenes_Rack_C>,pOpen_v3>
<Plate:<Container:FreeGenes_Rack_C>,pOpen_v3>
<Plate:<Container:FreeGenes_Rack_C>,pOpen_v3>
<Plate:<Container:FreeGenes_Rack_C>,pOpen_v3>
<Plate:<Container:FreeGenes_Rack_C>,pOpen_v3>
```
and we can see that we have one well per plate (one row in the export corresponds to a plate and well). If your expectation is for the export of a plateset map to only print the plate, we would still have 18 (but no information about wells).

```python
for plate in plates.all(): 
    ...:     print(plate.wells.count()) 
    ...:                                                                                                              
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
```

corresponding to 18 rows in the export, which seems correct based on what is in the database. Further, I can trace this back to the original data import from the old API, here is the plateset - notice the 18 plates (Keoni this is in the plateset-full.json from the data repo I sent you for import).

```
    {
        "description": "Plates with a single aliquot of pOpen_v3",
        "name": "pOpen_v3",
        "plates": [
            "642230d6-7c00-424f-a703-48aa43ef9974",
            "9b80fcd8-22af-4e6e-ac41-efe91fcc57cc",
            "72b9b5f3-88d5-4763-8afe-41aed16de4e0",
            "49f2bad2-d29a-40ae-a2a7-7e0bfcce2709",
            "2abe8dbf-5d6e-41ee-9e88-289937a2a1c7",
            "8f1da5da-34b2-4ef4-85ac-d346346db638",
            "97d8a81a-8849-494b-9c07-b78c9af0e600",
            "4087d4be-29be-4ecd-bd62-58118ad4443a",
            "68af348f-80ab-4ce4-a99a-613c7fc33ca2",
            "878f7ab7-3b03-47cf-a5c7-e1d56254aabd",
            "86397b4d-fa88-4f39-97bb-d02a676211ce",
            "25903331-33f9-413f-a71f-2c5131af6da0",
            "f679d2cb-6ee1-401e-95e5-16d08d701516",
            "45a3d655-387a-49a9-a60c-9ffca501b48f",
            "ad9b2284-a9b4-410b-9a5f-f9fb2f10ed2b",
            "4ea517cc-f0e6-4b86-8188-47ac354a799f",
            "5593df04-5c43-4d16-b460-250973c2b48b",
            "1824ac00-0de3-4fa0-821a-84b6b3e95236"
        ],
        "uuid": "37d895e3-eaf1-48f5-ad3d-9a7571e4434c"
    },
```

I can then load these raw files to verify that these relationships were already established:

```
with open("scripts/data/2019-09-03/plateset-full.json", "r") as filey:
    content = json.loads(filey.read())

plateset = content[0]
plateset

{'description': 'Plates with a single aliquot of pOpen_v3',
 'name': 'pOpen_v3',
 'plates': ['642230d6-7c00-424f-a703-48aa43ef9974',
  '9b80fcd8-22af-4e6e-ac41-efe91fcc57cc',
  '72b9b5f3-88d5-4763-8afe-41aed16de4e0',
  '49f2bad2-d29a-40ae-a2a7-7e0bfcce2709',
  '2abe8dbf-5d6e-41ee-9e88-289937a2a1c7',
  '8f1da5da-34b2-4ef4-85ac-d346346db638',
  '97d8a81a-8849-494b-9c07-b78c9af0e600',
  '4087d4be-29be-4ecd-bd62-58118ad4443a',
  '68af348f-80ab-4ce4-a99a-613c7fc33ca2',
  '878f7ab7-3b03-47cf-a5c7-e1d56254aabd',
  '86397b4d-fa88-4f39-97bb-d02a676211ce',
  '25903331-33f9-413f-a71f-2c5131af6da0',
  'f679d2cb-6ee1-401e-95e5-16d08d701516',
  '45a3d655-387a-49a9-a60c-9ffca501b48f',
  'ad9b2284-a9b4-410b-9a5f-f9fb2f10ed2b',
  '4ea517cc-f0e6-4b86-8188-47ac354a799f',
  '5593df04-5c43-4d16-b460-250973c2b48b',
  '1824ac00-0de3-4fa0-821a-84b6b3e95236'],
 'uuid': '37d895e3-eaf1-48f5-ad3d-9a7571e4434c'}
```
And now find the plates in the other plates export based on uuid:
```python
with open("scripts/data/2019-09-03/plates-full.json", "r") as filey:
    plates = json.loads(filey.read())

matches = dict()
for plate in plates:
    if plate['uuid'] in plateset['plates']:
        matches[plate['uuid']] = plate
```

And here they are, notice that each has one well, which is exactly what we see in that export. 

```python
{'642230d6-7c00-424f-a703-48aa43ef9974': {'container_uuid': 'ab74625e-d26e-4588-8e4b-4ba459aef5ad',
  'plate_form': 'standard96',
  'plate_name': 'pOpen_v3',
  'plate_type': 'glycerol_stock',
  'plate_vendor_id': None,
  'protocol_uuid': None,
  'status': 'Stocked',
  'thaw_count': 0,
  'uuid': '642230d6-7c00-424f-a703-48aa43ef9974',
  'wells': ['c6b5fa61-8c3b-4832-b3a3-b457f9690711']},
 '9b80fcd8-22af-4e6e-ac41-efe91fcc57cc': {'container_uuid': 'ab74625e-d26e-4588-8e4b-4ba459aef5ad',
  'plate_form': 'standard96',
  'plate_name': 'pOpen_v3',
  'plate_type': 'glycerol_stock',
  'plate_vendor_id': None,
  'protocol_uuid': None,
  'status': 'Stocked',
  'thaw_count': 0,
  'uuid': '9b80fcd8-22af-4e6e-ac41-efe91fcc57cc',
  'wells': ['9a6381aa-8a9e-419c-9aa3-6037a704277a']},
 '72b9b5f3-88d5-4763-8afe-41aed16de4e0': {'container_uuid': 'ab74625e-d26e-4588-8e4b-4ba459aef5ad',
  'plate_form': 'standard96',
  'plate_name': 'pOpen_v3',
  'plate_type': 'glycerol_stock',
  'plate_vendor_id': None,
  'protocol_uuid': None,
  'status': 'Stocked',
  'thaw_count': 0,
  'uuid': '72b9b5f3-88d5-4763-8afe-41aed16de4e0',
  'wells': ['9ebe7b21-9d09-457d-943a-e6d0df0c618a']},
 '49f2bad2-d29a-40ae-a2a7-7e0bfcce2709': {'container_uuid': 'ab74625e-d26e-4588-8e4b-4ba459aef5ad',
  'plate_form': 'standard96',
  'plate_name': 'pOpen_v3',
  'plate_type': 'glycerol_stock',
  'plate_vendor_id': None,
  'protocol_uuid': None,
  'status': 'Stocked',
  'thaw_count': 0,
  'uuid': '49f2bad2-d29a-40ae-a2a7-7e0bfcce2709',
  'wells': ['235d4e2e-51e9-4976-94af-c0d3544179e8']},
 '2abe8dbf-5d6e-41ee-9e88-289937a2a1c7': {'container_uuid': 'ab74625e-d26e-4588-8e4b-4ba459aef5ad',
  'plate_form': 'standard96',
  'plate_name': 'pOpen_v3',
  'plate_type': 'glycerol_stock',
  'plate_vendor_id': None,
  'protocol_uuid': None,
  'status': 'Stocked',
  'thaw_count': 0,
  'uuid': '2abe8dbf-5d6e-41ee-9e88-289937a2a1c7',
  'wells': ['01b4b86a-b9c7-4832-b5cd-af172bd8ab1e']},
 '8f1da5da-34b2-4ef4-85ac-d346346db638': {'container_uuid': 'ab74625e-d26e-4588-8e4b-4ba459aef5ad',
  'plate_form': 'standard96',
  'plate_name': 'pOpen_v3',
  'plate_type': 'glycerol_stock',
  'plate_vendor_id': None,
  'protocol_uuid': None,
  'status': 'Stocked',
  'thaw_count': 0,
  'uuid': '8f1da5da-34b2-4ef4-85ac-d346346db638',
  'wells': ['9c2efe40-dbb0-4c63-b964-69c135856fcd']},
 '97d8a81a-8849-494b-9c07-b78c9af0e600': {'container_uuid': 'ab74625e-d26e-4588-8e4b-4ba459aef5ad',
  'plate_form': 'standard96',
  'plate_name': 'pOpen_v3',
  'plate_type': 'glycerol_stock',
  'plate_vendor_id': None,
  'protocol_uuid': None,
  'status': 'Stocked',
  'thaw_count': 0,
  'uuid': '97d8a81a-8849-494b-9c07-b78c9af0e600',
  'wells': ['3d0aa1bf-eab5-421e-ac49-460f82fd9e53']},
 '4087d4be-29be-4ecd-bd62-58118ad4443a': {'container_uuid': 'ab74625e-d26e-4588-8e4b-4ba459aef5ad',
  'plate_form': 'standard96',
  'plate_name': 'pOpen_v3',
  'plate_type': 'glycerol_stock',
  'plate_vendor_id': None,
  'protocol_uuid': None,
  'status': 'Stocked',
  'thaw_count': 0,
  'uuid': '4087d4be-29be-4ecd-bd62-58118ad4443a',
  'wells': ['fcd38ab2-276c-43af-9272-574a209614b5']},
 '68af348f-80ab-4ce4-a99a-613c7fc33ca2': {'container_uuid': 'ab74625e-d26e-4588-8e4b-4ba459aef5ad',
  'plate_form': 'standard96',
  'plate_name': 'pOpen_v3',
  'plate_type': 'glycerol_stock',
  'plate_vendor_id': None,
  'protocol_uuid': None,
  'status': 'Stocked',
  'thaw_count': 0,
  'uuid': '68af348f-80ab-4ce4-a99a-613c7fc33ca2',
  'wells': ['575bbdd1-54e2-4efc-909b-e9fca0ce1da0']},
 '878f7ab7-3b03-47cf-a5c7-e1d56254aabd': {'container_uuid': 'ab74625e-d26e-4588-8e4b-4ba459aef5ad',
  'plate_form': 'standard96',
  'plate_name': 'pOpen_v3',
  'plate_type': 'glycerol_stock',
  'plate_vendor_id': None,
  'protocol_uuid': None,
  'status': 'Stocked',
  'thaw_count': 0,
  'uuid': '878f7ab7-3b03-47cf-a5c7-e1d56254aabd',
  'wells': ['569855c3-78dc-4e2e-807a-a91a81e035cb']},
 '86397b4d-fa88-4f39-97bb-d02a676211ce': {'container_uuid': 'ab74625e-d26e-4588-8e4b-4ba459aef5ad',
  'plate_form': 'standard96',
  'plate_name': 'pOpen_v3',
  'plate_type': 'glycerol_stock',
  'plate_vendor_id': None,
  'protocol_uuid': None,
  'status': 'Stocked',
  'thaw_count': 0,
  'uuid': '86397b4d-fa88-4f39-97bb-d02a676211ce',
  'wells': ['d8e2973c-ce5a-4eb1-9d5e-587496f4ae3c']},
 '25903331-33f9-413f-a71f-2c5131af6da0': {'container_uuid': 'ab74625e-d26e-4588-8e4b-4ba459aef5ad',
  'plate_form': 'standard96',
  'plate_name': 'pOpen_v3',
  'plate_type': 'glycerol_stock',
  'plate_vendor_id': None,
  'protocol_uuid': None,
  'status': 'Stocked',
  'thaw_count': 0,
  'uuid': '25903331-33f9-413f-a71f-2c5131af6da0',
  'wells': ['75d92bc3-dfd1-4787-b691-1756cf700180']},
 'f679d2cb-6ee1-401e-95e5-16d08d701516': {'container_uuid': 'ab74625e-d26e-4588-8e4b-4ba459aef5ad',
  'plate_form': 'standard96',
  'plate_name': 'pOpen_v3',
  'plate_type': 'glycerol_stock',
  'plate_vendor_id': None,
  'protocol_uuid': None,
  'status': 'Stocked',
  'thaw_count': 0,
  'uuid': 'f679d2cb-6ee1-401e-95e5-16d08d701516',
  'wells': ['e55e59eb-62ad-4f39-a242-6949a81f2434']},
 '45a3d655-387a-49a9-a60c-9ffca501b48f': {'container_uuid': 'ab74625e-d26e-4588-8e4b-4ba459aef5ad',
  'plate_form': 'standard96',
  'plate_name': 'pOpen_v3',
  'plate_type': 'glycerol_stock',
  'plate_vendor_id': None,
  'protocol_uuid': None,
  'status': 'Stocked',
  'thaw_count': 0,
  'uuid': '45a3d655-387a-49a9-a60c-9ffca501b48f',
  'wells': ['a94cb944-ef56-461d-8541-c61ee3e04111']},
 'ad9b2284-a9b4-410b-9a5f-f9fb2f10ed2b': {'container_uuid': 'ab74625e-d26e-4588-8e4b-4ba459aef5ad',
  'plate_form': 'standard96',
  'plate_name': 'pOpen_v3',
  'plate_type': 'glycerol_stock',
  'plate_vendor_id': None,
  'protocol_uuid': None,
  'status': 'Stocked',
  'thaw_count': 0,
  'uuid': 'ad9b2284-a9b4-410b-9a5f-f9fb2f10ed2b',
  'wells': ['8bfc53ab-350a-4652-ba5b-c864cf993b8c']},
 '4ea517cc-f0e6-4b86-8188-47ac354a799f': {'container_uuid': 'ab74625e-d26e-4588-8e4b-4ba459aef5ad',
  'plate_form': 'standard96',
  'plate_name': 'pOpen_v3',
  'plate_type': 'glycerol_stock',
  'plate_vendor_id': None,
  'protocol_uuid': None,
  'status': 'Stocked',
  'thaw_count': 0,
  'uuid': '4ea517cc-f0e6-4b86-8188-47ac354a799f',
  'wells': ['444ca9b8-4e3c-490c-bcab-51149009c300']},
 '5593df04-5c43-4d16-b460-250973c2b48b': {'container_uuid': 'ab74625e-d26e-4588-8e4b-4ba459aef5ad',
  'plate_form': 'standard96',
  'plate_name': 'pOpen_v3',
  'plate_type': 'glycerol_stock',
  'plate_vendor_id': None,
  'protocol_uuid': None,
  'status': 'Stocked',
  'thaw_count': 0,
  'uuid': '5593df04-5c43-4d16-b460-250973c2b48b',
  'wells': ['24473168-9f10-450f-8a07-1fa898638680']},
 '1824ac00-0de3-4fa0-821a-84b6b3e95236': {'container_uuid': 'ab74625e-d26e-4588-8e4b-4ba459aef5ad',
  'plate_form': 'standard96',
  'plate_name': 'pOpen_v3',
  'plate_type': 'glycerol_stock',
  'plate_vendor_id': None,
  'protocol_uuid': None,
  'status': 'Shipped',
  'thaw_count': 0,
  'uuid': '1824ac00-0de3-4fa0-821a-84b6b3e95236',
  'wells': ['83842394-0090-4128-8e41-171c1729fb84']}}
```
The reason they look the same is because:

 - the well address is the same (A1) across the plates (only one well)
 - the sample seems to be the same as well.

But the function / export is correctly showing 18 plates. The reason we don't see the rows as different is because there isn't anything entered for the last attribute, and the same seems to be the same? Specifically, here is what is added to each row (and how the loop works):

```python
# csv prep stuff up here!
...
    for plate in plates:
 
        # Add each well to the csv
        for well in plate.wells.all():

            # We get the part via the associated sample
            sample = well.sample_wells.first()
            writer.writerow([plate.name,
                             plate.plate_type,
                             well.address,
                             sample.part.name,
                             sample.part.gene_id,
                             sample.evidence,
                             sample.part.synthesized_sequence])

    return response
```

So I don't see how there could be just one row here given that 18 plates were imported from the original API. Could you help me to understand?

**Comment by vsoch** [comment]
Keoni are you saying that you are aware there are 18 (or more) plates, but this export should only show one plate (the first). Doesn't this defeat the purpose of a platemap export, to only show one plate? If we just need to show one, why does it need to export to csv at all? It's just a few attributes. What is the purpose of this export?

**Comment by Koeng101** [comment]
I am aware that there are 18 plates within that plate map, but to a naive viewer, they’re all identical to each other. When someone wants to get a distribution plate map, they basically just want to see what the plates they will get will look like.

The CSV export is so that the users can see what will be in each well of each plate (one from each plate map). Unless you’re really a power user, that’s all you need. We also have this as a feature on a per-plate level, but before we ship the plates, we don’t know which one we’ll give the end user.

**Comment by Koeng101** [comment]
TL;DR key is that each plate in a plateset is pretty much the exact same to an end user who wants a CSV export, so we just hide most of them

**Comment by vsoch** [comment]
okay - so you want the export for a plateset to only include one row, this is correct?

**Comment by vsoch** [comment]
Sorry that's not completely accurate - I meant to say, the export for a plateset to only include rows for one plate (all wells) so in the case of this example, it would just be one row.

**Comment by hverdonk** [comment]
Yes, that is correct.

**Comment by Koeng101** [comment]
Yep!

**Comment by vsoch** [comment]
okay all set! https://github.com/vsoch/freegenes/commit/857f844bacd57356df10f7857bd9cbfccba95e4e I added a title attribute (meaning it will work on mouseover) to the download csv buttons so that the user is aware that it's just showing a single representative plate. The download for distributions also does the same.
